### PR TITLE
Feature constraints style guide

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

Interested in getting this into develop before we take on more constraints

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [ ] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

**Review by aj-stein-gsa — COMMENTED** [review]
We need a rebase but this PR looks like a good start.

**Review by aj-stein-gsa — DISMISSED** [review]
Sometime today, can you rebase this PR and update the checklist accordingly please? Thanks.

**Comment by wandmagic** [comment]
> Sometime today, can you rebase this PR and update the checklist accordingly please? Thanks.

rebased ✅

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]

File: features/steps/fedramp_extensions_steps.ts [hunk 1]
```diff
@@ -665,4 +665,56 @@ Then("I should have both FAIL and PASS tests for constraint ID {string}", functi
...
+    console.log(filePath);
+    try {
+      console.log(filePath);
+      const [result, error] = await executeOscalCliCommand('metaschema', [
+        'validate',
+        filePath,
+        '-c',
+        styleGuidePath,
+        '--disable-schema-validation'
+      ]);
```
[the remark this answers is not in the document]
Comment (Gabeblis) [reply features/steps/fedramp_extensions_steps.ts:693]: We definitely can make 2 and 3 happen. As for 1, I'm not sure if this is pre-existing behavior, but the metaschema `validate` command is not producing any viable sarif. Also, something in the the new release is causing unwanted behavior when running the `validate` command. It isn't catching any errors and I verified that the same exact code works with 2.2, but not 2.3. Were there any significant changes that I'm missing? If not, then we're dealing with another bug.

[the remark this answers is not in the document]
Comment (aj-stein-gsa) [reply features/steps/fedramp_extensions_steps.ts:693]: I will take a look then.

**Comment by wandmagic** [comment]
ok added make lint, also added use 2.2.0 so we can use the older version until we sort out bugs in 2.3.0

File: features/steps/fedramp_extensions_steps.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aj-stein-gsa) [reply features/steps/fedramp_extensions_steps.ts:693]: Yeah something is not right here. It seems 2.2.0 works but 2.3.0 is not working I will work that bug report but we should figure that out over time. 

I see @wandmagic worked around that in https://github.com/GSA/fedramp-automation/pull/858/commits/3819f9c50b8e8e12abdf3b0d0eb82243fa6affd5 for now.

Comment (aj-stein-gsa) [inline features/steps/fedramp_extensions_steps.ts:693]: Now that we have rolled back to 2.2.0, can we please add something like this before final approval? Thanks!

```suggestion
  for (const file_name of constraint_files) {
    const filePath = join(constraintDir, file_name.trim());
    const styleGuideReportPath = join(styleGuideReportsDir, `${parse(filePath).name}.sarif`);
    console.log(filePath);
    try {
      console.log(filePath);
      const [result, error] = await executeOscalCliCommand('metaschema', [
        'validate',
        filePath,
        '-c',
        styleGuidePath,
        '-o',
        styleGuideReportPath,
        '--disable-schema-validation',
      ]);
```

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]

File: features/steps/fedramp_extensions_steps.ts [hunk 2]
```diff
@@ -665,4 +665,51 @@ Then("I should have both FAIL and PASS tests for constraint ID {string}", functi
...
+  const constraint_files = readdirSync(constraintDir).filter((file) => file.startsWith('fedramp') && file.endsWith('xml') );
+  const errors = [];
+
+  function filterOutBrackets(input) {
+    return input.replace(/\[.*?\]/g, '');
+  }
+
+  for (const file_name of constraint_files) {
+    const filePath = join(constraintDir, file_name.trim());
+    console.log(filePath);
```
Comment (Gabeblis) [inline features/steps/fedramp_extensions_steps.ts:684]: I may have left some debug logs in there 🙄 
```suggestion
```

File: features/steps/fedramp_extensions_steps.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ import { Exception, Log, Result } from "sarif";
 import { fileURLToPath } from "url";
 import { parseString } from "xml2js";
 import { promisify } from "util";
+import {formatSarifOutput} from 'oscal'
```
Comment (Gabeblis) [inline features/steps/fedramp_extensions_steps.ts:19]: After testing locally, the sarif output is not formatted correctly. The text appears as one large block.

Comment (wandmagic) [reply features/steps/fedramp_extensions_steps.ts:19]: fixed

**Review by aj-stein-gsa — APPROVED** [review]

**Comment by wandmagic** [comment]
This is ready for review again

**Review by aj-stein-gsa — APPROVED** [review]

**Comment by aj-stein-gsa** [comment]
Anything I need to know with the help wanted tag?

**Comment by wandmagic** [comment]
> Anything I need to know with the help wanted tag?

Just help reviewing

**Review by Gabeblis — APPROVED** [review]
Things are about to get REAL

**Comment by aj-stein-gsa** [comment]
> Things are about to get REAL

This made my evening, thanks to you and the whole crew who made it happen so quickly!

**Review by Rene2mt — APPROVED** [review]
Looks good!!!

**Pull request merged**